### PR TITLE
Make `createBackendPlugin` and `createBackendModule` return `BackendFeature`

### PR DESCRIPTION
**Pull request by freben**
Closes: https://github.com/backstage/backstage/issues/24920, https://github.com/backstage/backstage/issues/25122

Opting to keep the changesets, if you think that makes sense - the API effectively will be affected, but it's a matter of taste.

### CODE REVIEW

**Review by awanlin — APPROVED** [review]
Looks good for DevTools 👍

**Comment by freben** [comment]
There were some unexpected complexities when doing this - putting it on hold until next week

File: packages/backend-plugin-api/api-report.md [hunk 1]
```diff
@@ -208,7 +208,7 @@ export namespace coreServices {
 // @public
 export function createBackendModule(
   options: CreateBackendModuleOptions,
-): () => BackendFeature;
+): BackendFeature;
```
Comment (Rugvip) [inline packages/backend-plugin-api/api-report.md:211]: Would something like this work for now?

```ts
interface Derp extends BackendFeature {
  /** @deprecated nah */
  (): BackendFeature
}
```

Comment (freben) [reply packages/backend-plugin-api/api-report.md:211]: This is basically what I started in https://github.com/backstage/backstage/blob/62b2dab79c6ea148580e302ece49072cb04d0881/packages/backend-plugin-api/src/types.ts#L30

Inheriting so as not to pollute the original type is very clever.

But I ran into troubles at runtime with this. Haven't found the bug yet. Things end up with duplicate registrations.

Comment (Rugvip) [reply packages/backend-plugin-api/api-report.md:211]: Hmm, should be possible to roll this out as just a typescript change for now though, no runtime changes other than updated usage

Comment (freben) [reply packages/backend-plugin-api/api-report.md]: Done!

**Review by camilaibs — APPROVED** [review]
LGTM 💯

**Pull request merged**